frontend/controller/actions/group.js

```diff
@@ -651,23 +652,37 @@ export default (sbp('sbp/selectors/register', {
   }),
   ...encryptedAction('gi.actions/group/joinChatRoom', L('Failed to join chat channel.'), async function (sendMessage, params) {
     const rootState = sbp('chelonia/rootState')
-    const me = rootState.loggedIn.identityContractID
-    const memberID = params.data.memberID || me
+    const { identityContractID } = rootState.loggedIn
+    const memberID = params.data.memberID || identityContractID
+    const chatRoomID = params.data.chatRoomID
+    const groupContractID = params.contractID
 
     // If we are inviting someone else to join, we need to share the chatroom's keys
     // with them so that they are able to read messages and participate
-    if (memberID !== me && rootState[params.data.chatRoomID].attributes.privacyLevel === CHATROOM_PRIVACY_LEVEL.PRIVATE) {
+    if (memberID !== identityContractID && rootState[chatRoomID].attributes.privacyLevel === CHATROOM_PRIVACY_LEVEL.PRIVATE) {
       await sbp('gi.actions/out/shareVolatileKeys', {
         contractID: memberID,
         contractName: 'gi.contracts/identity',
-        subjectContractID: params.data.chatRoomID,
+        subjectContractID: chatRoomID,
         keyIds: '*'
       })
     }
 
-    return await sendMessage({
+    const response = await sendMessage({
       ...omit(params, ['options', 'action'])
     })
+
+    const switchChannelAfterJoined = (contractID: string) => {
+      if (contractID === chatRoomID) {
+        if (rootState[chatRoomID]?.members?.[identityContractID]) {
+          sbp('okTurtles.events/emit', JOINED_CHATROOM, { identityContractID, groupContractID, chatRoomID })
+          sbp('okTurtles.events/off', EVENT_HANDLED, switchChannelAfterJoined)
+        }
+      }
+    }
+    sbp('okTurtles.events/on', EVENT_HANDLED, switchChannelAfterJoined)
```

Comment:
This handler needs to be set up before calling `sendMessage`, not after, to make sure it is ready when the message comes back, otherwise it is possible the message will be processed before the handler is setup.

Reply:
Nice catch!